Sources/Kitura/String+Extensions.swift

```diff
@@ -44,32 +46,105 @@ extension String {
+        assert(aString == aString.lowercased(), "equalsLowercased() should be passed a lowercased string, not '\(aString)'")
+        return self.lowercased() == aString
+    #else
+        return self.caseInsensitiveCompare(aString) == .orderedSame
+    #endif
+    }
+}
+
+extension Substring {
     /// Splits a URL-encoded key and value pair (e.g. "foo=bar") into a tuple
     /// with corresponding "key" and "value" values, with the value being URL
     /// unencoded.
-    var keyAndDecodedValue: (key: String, value: String?) {
-        guard let range = self.range(of: "=") else {
+    var keyAndDecodedValue: (key: Substring, value: Substring?) {
+        #if swift(>=4.2)
+        guard let index = self.firstIndex(of: "=") else {
+            return (key: self, value: nil)
+        }
+        #else
+        guard let index = self.index(of: "=") else {
             return (key: self, value: nil)
         }
-        let key = String(self[..<range.lowerBound])
-        let value = String(self[range.upperBound...])
+        #endif
+        // substring up to index
+        let key = self[..<index]
+        // substring from index
+        var value = self[self.index(after: index)...]
+
+        // Faster way to replace '+' with ' ' that does not involve conversion to NSString
+        value.replaceCharacters("+", with: " ")
 
-        let valueReplacingPlus = value.replacingOccurrences(of: "+", with: " ")
-        let decodedValue = valueReplacingPlus.removingPercentEncoding
-        if decodedValue == nil {
-            Log.warning("Unable to decode query parameter \(key) (coded value: \(valueReplacingPlus)")
+        // Note: Foundation processing function
+        guard let decodedValue = value.removingPercentEncoding else {
+            Log.warning("Unable to decode query parameter \(key) (coded value: \(value)")
+            return (key: key, value: value)
         }
-        return (key: key, value: decodedValue ?? valueReplacingPlus)
+        return (key: key, value: Substring(decodedValue))
+    }
+
+    /// Finds and replaces all occurrences of a character with the provided substring
+    /// (eg. another character).
+    @inline(__always)
+    private mutating func replaceCharacters(_ src: Character, with dst: Substring) {
+        repeat {
+            #if swift(>=4.2)
+            guard let startIndex = self.firstIndex(of: src) else {
+                break
+            }
+            #else
+            guard let startIndex = self.index(of: src) else {
+                break
+            }
+            #endif
+            self.replaceSubrange(startIndex...startIndex, with: dst)
+        } while true
+    }
+
+    /// Trims space and tab characters from the start and end of a string. This is
+    /// equivalent to String.trimmingCharacters(in: .whitespaces) for ASCII strings.
+    /// This function should *not* be used for strings that may be padded by exotic
+    /// Unicode whitespaces.
+    func trimAsciiWhitespace() -> Substring {
```

Comment:
Re. `@inline`, although I applied it to the `replaceCharacters` function above it, on reflection I didn't give it a huge amount of consideration and certainly didn't benchmark the difference to establish it is necessary.  I should probably take the previous one out (the compiler may still choose to inline - I don't think there's anything preventing it from doing so).

In both of these cases, the function calls stdlib functions which could potentially be inlined into this one (which could be important since both functions contain a loop).  If I force the compiler to inline this one, my concern is that I could be interfering with other inlining decisions which would have the opposite effect from what I was hoping to achieve.  The compiler has the bigger picture and so I'd expect it to make an equal or better decision than my guesswork.